Sources/SQLite/SQLiteDatabaseChangePublisher.swift

```diff
@@ -0,0 +1,171 @@
+
+    private var updatedTables = Set<String>()
+    private let queue: OperationQueue
+
+    init(database: SQLiteDatabase) {
+        self.database = database
+        changeTrackerURL = database.changeTrackerURL
+
+        queue = OperationQueue()
+        queue.name = "app.shareup.sqlite.sqlitedatabasechangepublisher"
+        queue.maxConcurrentOperationCount = 1
+
+        super.init()
+
+        createHooks()
+        registerFilePresenter()
+        subscribeToChangeTracker()
+    }
+
+    deinit {
+        changeTrackerSubscription = nil
+        removeHooks()
+    }
+
+    func open() {
+        precondition(SQLiteQueue.isCurrentQueue)
+        downstreamSubject.send(.open)
+        createHooks()
+        registerFilePresenter()
+    }
+
+    func close() {
+        precondition(SQLiteQueue.isCurrentQueue)
+        removeHooks()
+        removeFilePresenter()
+        downstreamSubject.send(.close)
+    }
+
+    func receive<S: Subscriber>(
+        subscriber: S
+    ) where Failure == S.Failure, Output == S.Input {
+        downstreamSubject.receive(subscriber: subscriber)
+    }
+}
+
+private extension SQLiteDatabaseChangePublisher {
+    func createHooks() {
+        guard let database = self.database else { return }
+
+        SQLiteQueue.sync {
+            database.createUpdateHandler { [weak self] (table) in
+                guard let self = self else { return }
+                self.updatedTables.insert(table)
+            }
+            
+            database.createCommitHandler { [weak self] in
+                guard let self = self else { return }
+                let tables = self.updatedTables
+                self.updatedTables.removeAll()
+                self.notifyDownstreamSubscribersAsync(.updateTables(tables))
+                self.notifyOtherProcesses()
+            }
+
+            database.createRollbackHandler { [weak self] in
+                guard let self = self else { return }
+                self.updatedTables.removeAll()
+            }
+        }
+    }
+
+    func removeHooks() {
+        guard let database = self.database else { return }
+
+        SQLiteQueue.sync {
+            database.removeUpdateHandler()
+            database.removeCommitHandler()
+            database.removeRollbackHandler()
+        }
+    }
+}
+
+extension SQLiteDatabaseChangePublisher: NSFilePresenter {
+    var presentedItemURL: URL? { changeTrackerURL }
+    var presentedItemOperationQueue: OperationQueue { queue }
+
+    func presentedItemDidChange() {
+        notifyDownstreamSubscribersAsync(.crossProcessUpdate)
+    }
+
+    private func registerFilePresenter() {
+        guard let url = changeTrackerURL else { return }
+        touch(url)
+        NSFileCoordinator.addFilePresenter(self)
+    }
+
+    private func removeFilePresenter() {
+        NSFileCoordinator.removeFilePresenter(self)
+    }
+
+    private func subscribeToChangeTracker() {
+        let sub = changeTrackerSubject
+            .throttle(
```

Comment:
I don't want us to queue up hundreds of coordinated disk writes when making a bunch of changes to the database, such as when we're uploading or downloading a file. `throttle()` will make sure at least 1 second passes between each write. In the future, it might make sense to increase this number, but we'll try this for now.